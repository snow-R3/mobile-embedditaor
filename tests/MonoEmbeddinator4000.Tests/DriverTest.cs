using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using CppSharp;
using CppSharp.Generators;
using NUnit.Framework;

namespace MonoEmbeddinator4000.Tests
{
    /// <summary>
    /// A set of integration tests / approval tests verifying generated java code
    /// </summary>
    [TestFixture]
    public class DriverTest : TempFileTest
    {
        Project project;
        Options options;
        Driver driver;

        [SetUp]
        public override void SetUp()
        {
            base.SetUp();

            var outputDir = Path.Combine(Path.GetDirectoryName(GetType().Assembly.Location), "output");
            temp = Path.Combine(Path.GetTempPath(), "hello.dll");
            tempFiles = new List<string> { temp, outputDir };

            project = new Project();
            options = new Options
            {
                GeneratorKind = GeneratorKind.Java,
                CompileCode = true,
            };

            options.OutputDir =
                project.OutputPath = outputDir;
        }

        void RunDriver(string resourceFile)
        {
            var parameters = new CompilerParameters
            {
                OutputAssembly = temp,
            };
            AssemblyGenerator.CreateFromResource(resourceFile, parameters);

            project.Assemblies.Add(temp);

            driver = new Driver(project, options);
            Assert.IsTrue(driver.Run(), "Call to Driver.Run() failed!");
        }

        [Test]
        public void Empty()
        {
            driver = new Driver(new Project(), new Options
            {
                GeneratorKind = GeneratorKind.Java,
            });
            Assert.IsTrue(driver.Run()); //This runs but doesn't throw
        }

        [Test]
        public void HelloFiles()
        {
            RunDriver("Hello");

            var builder = new StringBuilder();
            foreach (var file in driver.Output.Files.Keys)
            {
                //NOTE: replace \ so this works on Windows
                builder.AppendLine(file.Replace('\\', '/'));
            }
            Approvals.Verify(builder.ToString());
        }

        [Test]
        public void Hello_java()
        {
            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path);
        }

        [Test]
        public void Native_hello_dll_java()
        {
            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.Last());
            Approvals.VerifyFile(path);
        }

        [Test]
        public void UpperCase_java()
        {
            RunDriver("HelloUpper");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path); //TODO: I don't know if "String wORLD()" is what we want
        }

        [Test]
        public void AssemblyWithDots()
        {
            temp = Path.Combine(Path.GetTempPath(), "hello.with.dots.dll");
            tempFiles.Add(temp);

            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path);
        }

        [Test]
        public void AssemblyWithUpperCase()
        {
            temp = Path.Combine(Path.GetTempPath(), "HELLO.dll");
            tempFiles.Add(temp);

            RunDriver("Hello");

            string path = Path.Combine(options.OutputDir, driver.Output.Files.Keys.First());
            Approvals.VerifyFile(path);
        }

        [Test, Category("Slow"), Platform("MacOSX")]
        public void JarFileContents()
        {
            options.Compilation.Platform = TargetPlatform.MacOS;
            options.GeneratorKind = GeneratorKind.C;
            RunDriver("Hello");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello");

            var aar = Path.Combine(options.OutputDir, "Hello.jar");
            Approvals.VerifyZipFile(aar);
        }

        [Test, Category("Slow")]
        public void AarFileContents()
        {
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            RunDriver("Hello");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello");

            var aar = Path.Combine(options.OutputDir, "Hello.aar");
            Approvals.VerifyZipFile(aar);
        }

        [Test, Category("Slow")]
        public void AarFileContentsDebug()
        {
            options.Compilation.Platform = TargetPlatform.Android;
            options.GeneratorKind = GeneratorKind.C;
            options.Compilation.DebugMode = true;
            RunDriver("Hello");
            options.GeneratorKind = GeneratorKind.Java;
            RunDriver("Hello");

            var aar = Path.Combine(options.OutputDir, "Hello.aar");
            Approvals.VerifyZipFile(aar);
        }
    }
}
