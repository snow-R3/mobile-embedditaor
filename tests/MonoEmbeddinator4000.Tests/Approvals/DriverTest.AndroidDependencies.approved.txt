META-INF/
META-INF/MANIFEST.MF
AndroidManifest.xml
assets/
assets/assemblies/
assets/assemblies/dependency.dll
assets/assemblies/hello.dll
assets/assemblies/Java.Interop.dll
assets/assemblies/Mono.Android.dll
assets/assemblies/Mono.Android.Export.dll
assets/assemblies/Mono.Security.dll
assets/assemblies/mscorlib.dll
assets/assemblies/Resource.designer.dll
assets/assemblies/System.Core.dll
assets/assemblies/System.dll
assets/assemblies/System.Runtime.Serialization.dll
assets/assemblies/System.Xml.dll
assets/test.txt
classes.jar
jni/
jni/arm64-v8a/
jni/arm64-v8a/libhello.so
jni/arm64-v8a/libjnidispatch.so
jni/arm64-v8a/libmonodroid.so
jni/arm64-v8a/libmonosgen-2.0.so
jni/arm64-v8a/libtest.so
jni/armeabi/
jni/armeabi/libhello.so
jni/armeabi/libjnidispatch.so
jni/armeabi/libmonodroid.so
jni/armeabi/libmonosgen-2.0.so
jni/armeabi/libtest.so
jni/armeabi-v7a/
jni/armeabi-v7a/libhello.so
jni/armeabi-v7a/libjnidispatch.so
jni/armeabi-v7a/libmonodroid.so
jni/armeabi-v7a/libmonosgen-2.0.so
jni/armeabi-v7a/libtest.so
jni/x86/
jni/x86/libhello.so
jni/x86/libjnidispatch.so
jni/x86/libmonodroid.so
jni/x86/libmonosgen-2.0.so
jni/x86/libtest.so
jni/x86_64/
jni/x86_64/libhello.so
jni/x86_64/libjnidispatch.so
jni/x86_64/libmonodroid.so
jni/x86_64/libmonosgen-2.0.so
jni/x86_64/libtest.so
R.txt
res/
res/layout/
res/layout/customview.xml
res/layout/hello.xml
res/layout/this_is_caps.xml
res/values/
res/values/attributes.xml
res/values/strings.xml
